packages/xstate-react/src/useMachine.ts

```diff
@@ -91,28 +65,63 @@ export function useMachine<TMachine extends AnyStateMachine>(
   getMachine: MaybeLazy<TMachine>,
   ...[options = {}]: RestParams<TMachine>
 ): UseMachineReturn<TMachine> {
-  const listener = useCallback((nextState: StateFrom<TMachine>) => {
-    // Only change the current state if:
-    // - the incoming state is the "live" initial state (since it might have new actors)
-    // - OR the incoming state actually changed.
-    //
-    // The "live" initial state will have .changed === undefined.
-    const initialStateChanged =
-      nextState.changed === undefined && Object.keys(nextState.children).length;
+  const service = useIdleInterpreter(getMachine, options as any);
 
-    if (nextState.changed || initialStateChanged) {
-      setState(nextState);
+  const getSnapshot = useCallback(() => {
+    if (service.status === InterpreterStatus.NotStarted) {
+      return (options.state
+        ? State.create(options.state)
+        : service.machine.initialState) as State<any, any, any, any, any>;
     }
-  }, []);
 
-  const service = useInterpret(getMachine as any, options as any, listener);
+    return service.state;
+  }, [service]);
 
-  const [state, setState] = useState(() => {
-    const { initialState } = service.machine;
-    return (options.state
-      ? State.create(options.state as any)
-      : initialState) as StateFrom<TMachine>;
-  });
+  const isEqual = useCallback(
```

Comment:
I would love to ditch this in the future, it's kinda hacky (I don't even check the `_prevState`) but this would require some changes in XState itself

Reply:
I guess I've already mentioned this in other comment from Jan: https://github.com/statelyai/xstate/pull/2939/files#r784176656